### ISSUE TITLE
Uniquely identifying a set of arguments

### ISSUE THREAD

**Issue by Technologicat** [Technologicat/unpythonic#26]
We currently identify a set of arguments (for caching purposes), passed in as `*args` and `**kwargs`, using the following `k` as the key to a cache dictionary:

```python
from operator import itemgetter
k = (args, tuple(sorted(kwargs.items(), key=itemgetter(0))))
```

(This functionality is only relevant for pure functions. Assume also that all relevant data is hashable.)

However, consider:

```python
def f(a):
    ...

f(42)    # --> args = (42,), kwargs = {}
f(a=42)  # --> args = (), kwargs = {a: 42}
```

The value of `k` will be different, even though exactly the same arguments are bound in the body of `f`! (Namely, in both cases, the formal parameter `a` has been bound to the value `42`.)

This affects at least `@fix`, `@memoize` and `@gmemoize`, and at least in an ideal world, should be considered a bug. Grep the codebase for "kwargs.items" to find all instances.

This is posted as **#discussion** and attached to the ∞ milestone mainly because at the moment of this writing, I don't yet have a plan.

So, if you're reading this and want to take it on: the question is, how to compute `k` such that the computed value (or its hash) uniquely represents a set of argument bindings? We can consider `f` as fixed; all use cases of this pattern are in decorators that already know which function they're being applied to.

Suggestions are welcome - it would be nice to fix this.

This gotcha was noticed by the author of the `wrapt` library, and is mentioned in [its documentation](https://wrapt.readthedocs.io/en/latest/decorators.html#processing-function-arguments). See the paragraphs beginning *"If needing to modify certain arguments..."* and *"You should not simply attempt to extract positional arguments..."*

**Comment by Technologicat** [comment]
We could probably add something like this... (just sketching, totally untested for now)

```python
# https://docs.python.org/3/library/inspect.html#inspect.Signature
# https://docs.python.org/3/library/inspect.html#inspect.Parameter
def resolve_bindings(f, *args, **kwargs):
    """Resolve parameter bindings established by `f` when called with the given args and kwargs.

    This is an inspection tool, which does not actually call `f`. This is useful for memoizers
    and other similar decorators that need a canonical representation of `f`'s parameter bindings.

    As an example, consider a simplistic memoizer::

        from operator import itemgetter

        def memoize(f):
            memo = {}
            def memoized(*args, **kwargs):
                k = (args, tuple(sorted(kwargs.items(), key=itemgetter(0))))
                if k in memo:
                    return memo[k]
                memo[k] = v = f(*args, **kwargs)
                return v
            return memoized

        @memoize
        def f(a):
            return 2 * a

        f(42)    # --> memoized sees args = (42,), kwargs = {}
        f(a=42)  # --> memoized sees args = (), kwargs = {"a": 42}

    Even though both calls bind {"a": 42} in the body of `f`, the memoizer sees the invocations
    differently, so its cache will miss.

    This is solved by `resolve_bindings`::

        from operator import itemgetter

        def memoize(f):
            memo = {}
            def memoized(*args, **kwargs):
                bindings = resolve_bindings(f, *args, **kwargs)  # --> {"a": 42} in either case
                k = tuple(bindings.items())
                if k in memo:
                    return memo[k]
                memo[k] = v = f(*args, **kwargs)
                return v
            return memoized

        @memoize
        def f(a):
            return 2 * a

        f(42)
        f(a=42)  # now the cache hits
    """
    f, _ = _getfunc(f)
    params = signature(f).parameters

    poskinds = set((Parameter.POSITIONAL_ONLY,
                    Parameter.POSITIONAL_OR_KEYWORD))
    kwkinds = set((Parameter.POSITIONAL_OR_KEYWORD,
                   Parameter.KEYWORD_ONLY))
    varkinds = set((Parameter.VAR_POSITIONAL,
                    Parameter.VAR_KEYWORD))

    index = {}
    nposparams = 0
    varpos = varkw = None
    for slot, param in enumerate(params.values()):
        if param.kind in poskinds:
            nposparams += 1
        if param.kind in kwkinds:
            index[param.name] = slot
        if param.kind == Parameter.VAR_POSITIONAL:
            varpos = param.name
        elif param.kind == Parameter.VAR_KEYWORD:
            varkw = param.name

    # https://docs.python.org/3/reference/compound_stmts.html#function-definitions
    # https://docs.python.org/3/reference/expressions.html#calls
    unassigned = object()  # gensym
    slots = [unassigned for _ in range(len(params))]  # yes, varparams too

    # fill from positional arguments
    for slot, (param, value) in enumerate(zip(params.values(), args)):
        if param.kind in varkinds:  # these are always last in the function def
            break
        slots[slot] = value

    if varpos:
        slots[index[varpos]] = []
    if varkw:
        slots[index[varkw]] = OrderedDict()
        vkdict = slots[index[varkw]]

    # gather excess positional arguments
    if len(args) > nposparams:
        if not varpos:
            raise TypeError("Too many positional arguments, and no *args to receive the rest")  # TODO: ape the standard message
        slots[index[varpos]] = args[nposparams:]

    # fill from keyword arguments
    for identifier, value in kwargs.items():
        if identifier in index:
            slot = index[identifier]
            if slots[slot] is unassigned:
                slots[slot] = value
            else:
                raise TypeError("Trying to assign the same parameter both positionally and by name")  # TODO: ape the standard message
        elif varkw:  # gather excess keyword arguments
            vkdict[identifier] = value
        else:
            raise TypeError("f has no such keyword parameter, and no **kwargs to receive the rest")  # TODO: ape the standard message

    # fill missing with defaults from function definition
    for slot, param in enumerate(params.values()):
        if slots[slot] is unassigned:
            if param.default is Parameter.empty:
                raise TypeError("Missing required parameter")  # TODO: ape the standard message
            slots[slot] = param.default

    # build the result dict
    bindings = OrderedDict()
    for param, value in zip(params.values(), slots):
        bindings[param.name] = value
    return bindings
```

**Comment by Technologicat** [comment]
**Issues cleanup**: Since now I have a plan and implementing it doesn't take much work, this is a low-hanging fruit. Let's try to have this in 0.14.2.

Note this won't fix kwargs handling in `curry`. That needs a better `partial` that tracks which kwargs are allowed and which have already been specified.

But this will fix spurious cache misses in `fix`, `memoize`, and `gmemoize`.

**Comment by Technologicat** [comment]
This was solved by inspecting the parameter bindings `f` would establish if it was called with the given args and kwargs. The rules are documented in:

https://docs.python.org/3/reference/compound_stmts.html#function-definitions
https://docs.python.org/3/reference/expressions.html#calls

so we just implement them independently of Python itself. The result is `arity.resolve_bindings`.

We also add a dynvar, `resolve_bindings_tuplify`, which can be set to `True` (via `with dyn.let`) to make `resolve_bindings` return a hashable result. The default is `False`, in which case it will return a `collections.OrderedDict` instead, which is nice for end-user use. But memoizers want a hashable result.

**Comment by Technologicat** [comment]
This could be migrated to use [`inspect.Signature.bind`](https://docs.python.org/3/library/inspect.html#inspect.Signature.bind), which *should* do the same thing.

**Comment by Technologicat** [comment]
Resolved in 4f2c0f6.
